File: GDTMobApp/GDTAds/GDTUnifiedNativeAd/UnifiedNativeAdViewController.m
//
//  UnifiedNativeAdViewController.m
//  GDTMobApp
//
//  Created by nimomeng on 2018/10/12.
//  Copyright © 2018 Tencent. All rights reserved.
//

#import "UnifiedNativeAdViewController.h"
#import "GDTUnifiedNativeAd.h"
#import "GDTAppDelegate.h"
#import "UnifiedNativeAdPortraitVideoViewController.h"
#import "UnifiedNativeAdFeedVideoTableViewController.h"
#import "UnifiedNativeAdVideoConfigView.h"


@interface UnifiedNativeAdViewController () <UITableViewDelegate, UITableViewDataSource>
@property (weak, nonatomic) IBOutlet UITextField *placementTextField;
@property (weak, nonatomic) IBOutlet UITableView *tableView;

@property (assign, nonatomic)  float minVideoDurationSliderValue;
@property (assign, nonatomic)  float maxVideoDurationSliderValue;
@property (nonatomic, strong) GDTVideoConfig *videoConfig;
@property (nonatomic, strong) NSArray *demoArray;

//自渲染增加广告位选择功能
@property (nonatomic, strong) UIAlertController *advStyleAlertController;

@property (nonatomic, strong) NSArray *advTypeTextArray;

@property (nonatomic, strong) UIButton *changAdvStyleButton;

@property (nonatomic, copy) NSString *token;
@property (nonatomic, assign) BOOL useToken;

@end

@implementation UnifiedNativeAdViewController

static NSString *ONE_BIGPHOTO_STR = @"2010198333498040";

static NSString *ONE_SMALL_PHOTO_STR = @"6060695393196051";

static NSString *ONE_WIDER_PHOTO_STR = @"4030598303592073";

static NSString *THREE_PHOTO_OR_ONE_PHOTO_STR = @"2000566593234845";

static NSString *ONE_BIG_WIDER_PHOTO_STR = @"1050394363190036";

//video
static NSString *ONE_LONGER_VIDEO_STR = @"3050349752532954";

static NSString *THREE_VIDEO_OR_ONE_VIDEO_STR = @"5040995343496165";

static NSString *ONE_WIDER_VIDEO_STR = @"7000593393992138";

static NSString *ONE_WIDER_OR_LONGER_VIDEO_STR = @"2040594303892119";


static NSInteger ADVTYPE_COUNT = 6;

- (void)viewDidLoad {
    [super viewDidLoad];
    
    [self initVideoConfig];
    self.advTypeTextArray = @[
                                  @[@"主图尺寸1280×720（纯图）",@"UnifiedNativeAdFeedImageViewController"],
                                  @[@"三小图或一张1280*720（纯图）",@"UnifiedNativeAdFeedImageViewController"],
                                  @[@"主图尺寸1080×1920或800×1200（纯图）", @"UnifiedNativeAdFeedImageViewController"],
                                  @[@"主图尺寸1280×720（视频）", @"UnifiedNativeAdFeedVideoTableViewController"],
                                  @[@"主图尺寸720×1280（视频）", @"UnifiedNativeAdFeedVideoTableViewController"],
                                  @[@"流量分配", @"UnifiedNativeAdFeedImageViewController"],];
    self.demoArray = @[
                        @[@"图片Feed", @"UnifiedNativeAdFeedImageViewController"],
                        @[@"视频Feed", @"UnifiedNativeAdFeedVideoTableViewController"],
                        @[@"沉浸式视频流", @"UnifiedNativeAdPortraitVideoViewController"],
                        @[@"视频信息流", @"UnifiedNativeAdPortraitFeedViewController"],
                        @[@"视频贴片广告", @"UnifiedNativePreVideoViewController"],
                        ];
    
    self.tableView.delegate = self;
    self.tableView.dataSource = self;
    
}

- (IBAction)selectADVStyle:(id)sender {
    self.advStyleAlertController = [UIAlertController alertControllerWithTitle:@"请选择需要的广告样式" message:nil preferredStyle:UIAlertControllerStyleActionSheet];
    NSArray *advTypePosIDArray = @[
                                  ONE_BIGPHOTO_STR,
                                  THREE_PHOTO_OR_ONE_PHOTO_STR,
                                  ONE_BIG_WIDER_PHOTO_STR,
                                  THREE_VIDEO_OR_ONE_VIDEO_STR,
                                  ONE_WIDER_VIDEO_STR,
                                  [self mediationId],
    ];
        for (NSInteger i = 0; i < ADVTYPE_COUNT; i++) {
            if (i == 10) {
                
            }
            
            UIAlertAction *advTypeAction = [UIAlertAction actionWithTitle:self.advTypeTextArray[i][0]
                                                                    style:UIAlertActionStyleDefault
                                                                  handler:^(UIAlertAction * _Nonnull action) {
                self.placementTextField.placeholder = advTypePosIDArray[i];
            }];
            [self.advStyleAlertController addAction:advTypeAction];
        }
        [self presentViewController:self.advStyleAlertController
                           animated:YES
                         completion:^{
            [self clickBackToMainView];
        }];
}

- (NSString *)mediationId {
    return @"101367";
}

- (void)clickBackToMainView {
    NSArray *arrayViews = [UIApplication sharedApplication].keyWindow.subviews;
    UIView *backToMainView = [[UIView alloc] init];
    for (int i = 1; i < arrayViews.count; i++) {
        NSString *viewNameStr = [NSString stringWithFormat:@"%s",object_getClassName(arrayViews[i])];
        if ([viewNameStr isEqualToString:@"UITransitionView"]) {
            backToMainView = [arrayViews[i] subviews][0];
            break;
        }
    }
//    UIView *backToMainView = [arrayViews.lastObject subviews][0];
    backToMainView.userInteractionEnabled = YES;
    UITapGestureRecognizer *backTap = [[UITapGestureRecognizer alloc] initWithTarget:self action:@selector(backTap)];
    [backToMainView addGestureRecognizer:backTap];
}

- (void)backTap {
    [self.advStyleAlertController dismissViewControllerAnimated:YES completion:nil];
}

- (void)touchesBegan:(NSSet<UITouch *> *)touches withEvent:(UIEvent *)event
{
    [super touchesBegan:touches withEvent:event];
    [self.view endEditing:YES];
}

- (void)jumpToAnotherView{
    [self.navigationItem.rightBarButtonItem setEnabled:NO];
    UnifiedNativeAdVideoConfigView *configView = [[UnifiedNativeAdVideoConfigView alloc] initWithFrame:CGRectMake(0, 0, self.view.frame.size.width, self.view.frame.size.height) theVideoConfig:self.videoConfig];
    __weak __typeof(self) weakSelf = self;
    configView.placementId = self.placementTextField.text.length > 0? self.placementTextField.text: self.placementTextField.placeholder;
    configView.minVideoDurationSlider.value = self.minVideoDurationSliderValue;
    configView.maxVideoDurationSlider.value = self.maxVideoDurationSliderValue;
    configView.useTokenSwitch.on = self.useToken;
    configView.tokenLabel.text = self.token;
    configView.callBackBlock = ^(GDTVideoConfig *videoConfig,float MinVideoDurationSliderValue,float MaxVideoDurationSliderValue,BOOL navigationRightButtonIsenabled){
        weakSelf.videoConfig = videoConfig;
        [weakSelf.navigationItem.rightBarButtonItem setEnabled:navigationRightButtonIsenabled];
        weakSelf.minVideoDurationSliderValue = MinVideoDurationSliderValue;
        weakSelf.maxVideoDurationSliderValue = MaxVideoDurationSliderValue;
    };
    configView.tokenBlock = ^(BOOL useToken, NSString * _Nonnull token) {
        weakSelf.useToken = useToken;
        weakSelf.token = token;
    };
    [configView showInView:self.view];
}
- (void)initVideoConfig
{
    self.videoConfig = [[GDTVideoConfig alloc] init];
    self.minVideoDurationSliderValue = 5;
    self.maxVideoDurationSliderValue = 60;
    self.navigationItem.rightBarButtonItem = [[UIBarButtonItem alloc] initWithTitle:@"更多视频配置" style:UIBarButtonItemStylePlain target:self action:@selector(jumpToAnotherView)];
}

#pragma mark - UITableView
- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section
{
    return self.demoArray.count;
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath
{
    static NSString *SimpleTableIdentifier = @"SimpleTableIdentifier";
    UITableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:SimpleTableIdentifier];
    if (cell == nil) {
        cell = [[UITableViewCell alloc] initWithStyle:UITableViewCellStyleDefault reuseIdentifier:SimpleTableIdentifier];
    }
    cell.textLabel.text = self.demoArray[indexPath.row][0];
    return cell;
}

- (void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath
{
    [tableView deselectRowAtIndexPath:indexPath animated:YES];
    NSString *className = self.demoArray[indexPath.row][1];
    UnifiedNativeAdBaseViewController *vc = [[NSClassFromString(className) alloc] init];
    vc.appId = kGDTMobSDKAppId;
    vc.placementId = self.placementTextField.text.length > 0? self.placementTextField.text: self.placementTextField.placeholder;
//    vc.maxVideoDuration = (NSInteger)self.maxVideoDurationSlider.value;
//    GDTVideoConfig *videoConfig = [[GDTVideoConfig alloc] init];
//    videoConfig.autoPlayPolicy = [self.autoPlayPolicyTextField.text integerValue];
//    videoConfig.videoMuted = self.shouldMuteOnVideoSwitch.isOn;
//    videoConfig.detailPageEnable = self.videoDetailPageEnableSwitch.isOn;
//    videoConfig.autoResumeEnable = self.autoResumeEnableSwitch.isOn;
//    videoConfig.userControlEnable = self.userControlEnableSwitch.isOn;
//    videoConfig.progressViewEnable = self.progressViewEnableSwitch.isOn;
//    videoConfig.coverImageEnable = self.coverImageEnableSwitch.isOn;
//    vc.videoConfig = videoConfig;
    vc.minVideoDuration = (NSInteger)self.minVideoDurationSliderValue;
    vc.maxVideoDuration = (NSInteger)self.maxVideoDurationSliderValue;
    vc.videoConfig = self.videoConfig;
    vc.token = self.token;
    vc.useToken = self.useToken;
    [self.navigationController pushViewController:vc animated:YES];
}


@end

File: GDTMobSample-Swift/GDTAdTypes/UnifiedNativeAdViewController/UnifiedNativeAdViewController.swift
//
//  UnifiedNativeAdViewController.swift
//  GDTMobSample-Swift
//
//  Created by nimomeng on 2018/11/28.
//  Copyright © 2018 Tencent. All rights reserved.
//

import UIKit

class UnifiedNativeAdViewController: UIViewController{
    
    @IBOutlet weak var placementTextField: UITextField!
    @IBOutlet weak var autoPlayPolicyTextField: UITextField!
    @IBOutlet weak var shouldMuteOnVideoSwitch: UISwitch!
    @IBOutlet weak var videoDetailPageEnableSwitch: UISwitch!
    @IBOutlet weak var userControlEnableSwitch: UISwitch!
    @IBOutlet weak var autoResumeEnableSwitch: UISwitch!
    @IBOutlet weak var progressViewEnableSwitch: UISwitch!
    @IBOutlet weak var coverImageEnableSwitch: UISwitch!
    @IBOutlet weak var maxVideoDurationLabel: UILabel!
    @IBOutlet weak var maxVideoDurationSlider: UISlider!
    @IBOutlet weak var tableView: UITableView!
    
    var videoConfig:GDTVideoConfig!
    private var demoArray =
        ["图片Feed",
        "视频Feed",
        "竖版全屏视频",
        "竖版Feed视频",
        "视频贴片广告"]
    
    
    override func viewDidLoad() {
        super.viewDidLoad()
        self.initVideoConfig()
        self.tableView.delegate = self
        self.tableView.dataSource = self
        self.maxVideoDurationSlider.addTarget(self, action: #selector(sliderMaxVideoDurationChanged), for: .valueChanged)
        self.maxVideoDurationLabel.text = String(format:"视频最大长: %.f",maxVideoDurationSlider.value)
    }

    override func touchesBegan(_ touches: Set<UITouch>, with event: UIEvent?) {
        super.touchesBegan(touches, with: event)
        self.view.endEditing(true)
    }
    
    @objc func sliderMaxVideoDurationChanged(){
        self.maxVideoDurationLabel.text = String(format:"视频最大长: %.f",maxVideoDurationSlider.value)
    }
    
    func initVideoConfig(){
        self.videoConfig = GDTVideoConfig()
        self.autoPlayPolicyTextField.text = String(format:" %d",self.videoConfig.autoPlayPolicy.rawValue)
        self.shouldMuteOnVideoSwitch.isOn = self.videoConfig.videoMuted
        self.videoDetailPageEnableSwitch.isOn = self.videoConfig.detailPageEnable
        self.userControlEnableSwitch.isOn = self.videoConfig.userControlEnable
        self.autoResumeEnableSwitch.isOn = self.videoConfig.autoResumeEnable
        self.progressViewEnableSwitch.isOn = self.videoConfig.progressViewEnable
        self.coverImageEnableSwitch.isOn = self.videoConfig.coverImageEnable
    }
}
// tableVIewDelegate
extension UnifiedNativeAdViewController:UITableViewDataSource,UITableViewDelegate{
    func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        return self.demoArray.count
    }
    
        

    
    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        let SimpleTableIdentifier = "SimpleTableIdentifier"
        let cell = UITableViewCell(style: .default, reuseIdentifier: SimpleTableIdentifier)
        cell.textLabel?.text = self.demoArray[indexPath.row]
        return cell
    }
    
    func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
        tableView.deselectRow(at: indexPath, animated: true)
        let vcType = demoArray[indexPath.row]
        let vc:UnifiedNativeAdBaseViewController?
        
        
        switch vcType {
        case "图片Feed":
            vc = UnifiedNativeAdFeedImageViewController()
        case "视频Feed":
            vc = UnifiedNativeAdFeedVideoTableViewController()
        case "竖版全屏视频":
            vc = UnifiedNativeAdPortraitVideoViewController()
        case "竖版Feed视频":
            vc = UnifiedNativeAdPortraitFeedViewController()
        case "视频贴片广告":
            vc = UnifiedNativePreVideoViewController()
        default:
            vc = nil
        }
        
        vc?.appId = Constant.kGDTMobSDKAppId
        vc?.placementId = (self.placementTextField.text!.count > 0 ? self.placementTextField!.text:self.placementTextField!.placeholder)!
        vc?.maxVideoDuration = Int(self.maxVideoDurationSlider.value)

        let videoConfig:GDTVideoConfig = GDTVideoConfig()
        videoConfig.videoMuted = self.shouldMuteOnVideoSwitch.isOn;
        videoConfig.detailPageEnable = self.videoDetailPageEnableSwitch.isOn;
        videoConfig.autoResumeEnable = self.autoResumeEnableSwitch.isOn;
        videoConfig.userControlEnable = self.userControlEnableSwitch.isOn;
        videoConfig.progressViewEnable = self.progressViewEnableSwitch.isOn;
        videoConfig.coverImageEnable = self.coverImageEnableSwitch.isOn;
        vc?.videoConfig = videoConfig
        
        if vc is UnifiedNativeAdPortraitVideoViewController {
            self.present(vc!, animated: true, completion: nil)
        }else {
            self.navigationController?.pushViewController(vc!, animated: true)
        }
        
        
    }
}

